File: algorithm-java/src/main/java/com/algorithm/chapterone/normal/third/Bag.java
package com.algorithm.chapterone.normal.third;

import java.util.Iterator;

/**
 * Bag implement by linked table
 * 
 * @author chenlixin at 2017年2月21日 上午11:34:58
 */
public class Bag<Item> implements Iterable<Item> {

	/**
	 * first node
	 */
	private Node first;
	/**
	 * length of bag
	 */
	private int len;

	@Override
	public Iterator<Item> iterator() {
		return new ListIterator();
	}

	/**
	 * constructor Bag
	 */
	public Bag() {
	}

	/**
	 * add element
	 * 
	 * @param item
	 */
	public void add(Item item) {
		Node oldFirst = first;
		first = new Node();
		first.item = item;
		first.next = oldFirst;
		len++;
	}

	/**
	 * Returns true if this list contains no elements.
	 * 
	 * @return
	 */
	public boolean isEmpty() {
		// return len == 0;
		return first == null;
	}

	/**
	 * size
	 * 
	 * @return size
	 */
	public Integer size() {
		return len;
	}

	/**
	 * Node
	 */
	private class Node {
		/**
		 * current node data
		 */
		private Item item;
		/**
		 * next node
		 */
		private Node next;
	}

	/**
	 * Iterator
	 */
	private class ListIterator implements Iterator<Item> {

		/**
		 * current node
		 */
		private Node current = first;

		@Override
		public boolean hasNext() {
			return current == null;
		}

		@Override
		public Item next() {
			Item item = current.item;
			current = current.next;
			return item;
		}

		@Override
		public void remove() {
		}
	}
}

File: algorithm-go/src/algorithm/chapterone/bag.go
package chapterone

// data type
type bagData interface{}

// Bag's methods
type Bag interface {
	Size() int
	IsEmpty() bool
	Add(data bagData)
}

// type bagNode
type bagNode struct {
	data bagData
	next *bagNode
}

var (
	length int
)

// init bag
func InitBag() *bagNode {
	return &bagNode{}
}

// add
func (head *bagNode) Add(data bagData) {
	oldFirst := head
	newHead := bagNode{data: data}
	newHead.next = oldFirst
	length++
}

// size
func (head *bagNode) Size() int {
	return length
}

// isEmpty
func (head *bagNode) IsEmpty() bool {
	first := head
	return first == nil
}
